export/kafka.go

```diff
@@ -0,0 +1,153 @@
+		panic(err)
+	}
+
+	kafka := &Kafka{
+		config:      config,
+		kafkaConfig: kafkaConfig,
+		topicInfo:   topicInfo,
+		producer:    producer,
+	}
+
+	agent := newExporterAgent(kafka)
+	if err := agent.Start(kafka.config.ReportingPeriodms); err != nil {
+		panic(err)
+	}
+
+	return kafka, agent
+}
+
+func createTopic(topicInfo TopicConfig, kafkaConfig *kafka.ConfigMap) {
+	adminClient, err := kafka.NewAdminClient(kafkaConfig)
+	if err != nil {
+		fmt.Printf("Failed to create Admin client: %s\n", err)
+		os.Exit(1)
+	}
+
+	// Contexts are used to abort or limit the amount of time
+	// the Admin call blocks waiting for a result.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create topics on cluster.
+	// Set Admin options to wait for the operation to finish (or at most 60s)
+	maxDuration, err := time.ParseDuration("60s")
+	if err != nil {
+		panic("time.ParseDuration(60s)")
+	}
+
+	if topicInfo.NumPartitions != 0 && topicInfo.NumReplicas != 0 {
+		results, err := adminClient.CreateTopics(ctx,
+			[]kafka.TopicSpecification{{
+				Topic:             topicInfo.Topic,
+				NumPartitions:     topicInfo.NumPartitions,
+				ReplicationFactor: topicInfo.NumReplicas,
+			}},
+			kafka.SetAdminOperationTimeout(maxDuration))
+
+		if err != nil {
+			fmt.Printf("Problem during the topic creation: %v\n", err)
+			os.Exit(1)
+		}
+
+		// Check for specific topic errors
+		for _, result := range results {
+			if result.Error.Code() != kafka.ErrNoError &&
+				result.Error.Code() != kafka.ErrTopicAlreadyExists {
+				fmt.Printf("Topic creation failed for %s: %v",
+					result.Topic, result.Error.String())
+				os.Exit(1)
+			}
+		}
+	}
+
+	adminClient.Close()
+
+}
+
+// Stop closes the Kafka producer.
+func (e *Kafka) Stop() {
+	defer e.producer.Close()
+}
+
+// ExportMetrics converts the metrics to a metrics service request protobuf and
+// makes a POST request with that payload to a Kafka broker.
+func (e Kafka) ExportMetrics(ctx context.Context, data []*metricdata.Metric) error {
+	go handleEvents(e.producer.Events())
+
+	for _, d := range data {
+		if matched, _ := regexp.Match(e.config.IncludeFilter, []byte(d.Descriptor.Name)); matched {
+			d.Resource, _ = resource.FromEnv(ctx)
+			metricsRequestpb := metricToProto(d)
+			payload, err := proto.Marshal(metricsRequestpb)
+			if err != nil {
+				log.Fatal("Marshalling Error: ", err)
+			}
+
+			err = e.producer.Produce(&kafka.Message{
+				TopicPartition: kafka.TopicPartition{
+					Topic:     &e.topicInfo.Topic,
+					Partition: kafka.PartitionAny, // or e.partition?
+				},
+				Value: payload,
+			}, nil)
+
+			if err != nil {
+				log.Fatal("Error sending message with Producer: ", err)
+			}
+		}
+	}
+
+	e.producer.Flush(15 * 1000)
+	return nil
+}
+
+func handleEvents(events chan kafka.Event) {
+	for e := range events {
+		switch ev := e.(type) {
+		case *kafka.Message:
+			if ev.TopicPartition.Error != nil {
+				fmt.Printf("Delivery failed: %v\n", ev.TopicPartition)
```

Comment:
we may want to do some rate limitting on error messages. What we've seen in Java is that when sending a batch of 1000 metrics, if the whole batch fails then you end up with 1,000 messages in the logs. See: https://github.com/confluentinc/ce-kafka/blob/f24c695758a2628759659f22cfd8efd387f943c4/ce-metrics/src/main/java/io/confluent/telemetry/exporter/kafka/KafkaExporter.java#L165-L173